pandas/io/parsers.py

```diff
@@ -975,17 +976,27 @@ def _is_index_col(col):
     return col is not None and col is not False
 
 
+def _evaluate_usecols(usecols, names):
+    if callable(usecols):
+        return set([i for i, name in enumerate(names)
+                    if com._apply_if_callable(usecols, name)])
+    else:
+        return usecols
+
+
 def _validate_usecols_arg(usecols):
     """
     Check whether or not the 'usecols' parameter
     contains all integers (column selection by index)
     or strings (column by name). Raises a ValueError
     if that is not the case.
     """
-    msg = ("The elements of 'usecols' must "
-           "either be all strings, all unicode, or all integers")
+    msg = ("'usecols' must either be all strings, all unicode, "
+           "all integers or callable")
 
     if usecols is not None:
+        if callable(usecols):
+            return usecols
```

Comment:
you can just in-line ``_evaluate_usecols`` here, then change ``_vaidate_usecols`` to return the usecols and set it